_posts/2022-03-21-getting-started-with-protobufs-for-typescript.md

```diff
@@ -0,0 +1,259 @@
+5. [Do something with the serialised message](#do-something-with-the-serialised-message)
+6. [De-serialise the message](#de-serialise-the-message)
+
+## Define the Data Structure for Your Messages
+
+First thing to realise about Protobufs are they are just an efficient (small data size) way of encoding messages. Next thing to realise is that these messages have a pre-defined data structure. You need to know that data structure at both ends, to serialize the message before sending and then de-serialize the message back into something useable when it has been received. This is what the `.proto` files are for, they define the data structure of the messages in a language independent format.
+
+In this example, we are going to keep it really simple because the [proto3 data structure](https://developers.google.com/protocol-buffers/docs/proto3) is not the interesting bit for this post. I am also shamelessly basing this on the [Google Protobuf Tutorials](https://developers.google.com/protocol-buffers/docs/tutorials) because I am sure you have already been there, hence using the Address Book example.
+
+So we are going to create an Address Book message that we want to transmit using Protobuf. This message contains People and People have contact details. For example, a JSON object for this message would look something like:
+
+```js
+const myAddressBook = {
+  people: [
+    {
+      name: "Joe Blogs",
+      phones: [
+        {
+          phoneNumber: "0123456789",
+          phoneType: "MOBILE",
+        },
+      ],
+    },
+    {
+      name: "Jane Smith",
+      phones: [
+        {
+          phoneNumber: "0987654321",
+          phoneType: "HOME",
+        },
+      ],
+    },
+  ],
+};
+```
+
+This structure becomes the following `addressBook.proto` file:
+
+```js
+syntax = "proto3";
+
+package tutorial;
+
+message Person {
+  optional string name = 1;
+
+  enum PhoneType {
+    MOBILE = 0;
+    HOME = 1;
+    WORK = 2;
+  }
+
+  message PhoneNumber {
+    optional string phoneNumber = 1;
+    optional PhoneType phoneType = 2;
+  }
+
+  repeated PhoneNumber phones = 2;
+}
+
+message AddressBook {
+  repeated Person people = 1;
+}
+```
+
+## Compile the TypeScript You Will Use in Your Code
+
+This is the first bit that didn’t quite make sense, there's some magic that's needed to turn the `.proto` files into language specific code you can use within your project. At this point in the journey, all you need to know is how to generate this output for your desired language.
+
+- For this you need the `protoc` compiler. Download and install the [prebuilt binary](https://github.com/protocolbuffers/protobuf/releases)
+  - For Mac and Linux you just have to copy the `protoc` file into a location on your `$PATH`. On my Mac I copied this into `/usr/local/bin/`.
+
+However, this only generates code for [some languages](https://github.com/protocolbuffers/protobuf#protobuf-runtime-installation). From that list we're interested in JavaScript, which is included.
```

Comment:
'which is included' is redundant because you have already said the list contains javascript.

Reply:
removed